File: server/src/medicscribe_server/ws/protocol.py
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field


class StartMessage(BaseModel):
    type: Literal["start"]
    # session_id is used to build the on-disk WAV path — constrain to a safe slug
    # (no '/', '.', whitespace) so it can never traverse out of audio_dir (CWE-22).
    # Browser/Android send a UUID, which matches this.
    session_id: str = Field(pattern=r"^[A-Za-z0-9_-]{1,128}$")
    template: str = "soap_v1"
    languages: list[str] = Field(default_factory=lambda: ["en", "zh", "ms", "ta"])


class StopMessage(BaseModel):
    type: Literal["stop"]


ClientMessage = Annotated[
    StartMessage | StopMessage,
    Field(discriminator="type"),
]


class AckMessage(BaseModel):
    type: Literal["ack"] = "ack"
    session_id: str


class TranscriptPartial(BaseModel):
    type: Literal["transcript_partial"] = "transcript_partial"
    text: str
    speaker: str | None = None
    lang: str | None = None
    t: float


class TranscriptFinal(BaseModel):
    type: Literal["transcript_final"] = "transcript_final"
    text: str
    speaker: str | None = None
    lang: str | None = None
    t: float


class AudioDeleted(BaseModel):
    type: Literal["audio_deleted"] = "audio_deleted"
    session_id: str


class NoteProgress(BaseModel):
    type: Literal["note_progress"] = "note_progress"
    stage: str
    pct: int


class NoteDone(BaseModel):
    type: Literal["note_done"] = "note_done"
    note: dict[str, Any]
    raw_transcript: str


class ErrorMessage(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str


ServerMessage = Annotated[
    AckMessage
    | TranscriptPartial
    | TranscriptFinal
    | AudioDeleted
    | NoteProgress
    | NoteDone
    | ErrorMessage,
    Field(discriminator="type"),
]

File: app/tablet/src/main/kotlin/com/medicscribe/net/Protocol.kt
package com.medicscribe.net

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject

@Serializable
sealed class ClientMessage {
    @Serializable
    @SerialName("start")
    data class Start(
        @SerialName("session_id") val sessionId: String,
        val template: String = "soap_v1",
        val languages: List<String> = listOf("en", "zh", "ms", "ta"),
    ) : ClientMessage()

    @Serializable
    @SerialName("stop")
    data object Stop : ClientMessage()
}

@Serializable
sealed class ServerMessage {
    @Serializable
    @SerialName("ack")
    data class Ack(
        @SerialName("session_id") val sessionId: String,
    ) : ServerMessage()

    @Serializable
    @SerialName("transcript_partial")
    data class TranscriptPartial(
        val text: String,
        val speaker: String? = null,
        val lang: String? = null,
        val t: Float,
    ) : ServerMessage()

    @Serializable
    @SerialName("transcript_final")
    data class TranscriptFinal(
        val text: String,
        val speaker: String? = null,
        val lang: String? = null,
        val t: Float,
    ) : ServerMessage()

    @Serializable
    @SerialName("note_progress")
    data class NoteProgress(
        val stage: String,
        val pct: Int,
    ) : ServerMessage()

    @Serializable
    @SerialName("note_done")
    data class NoteDone(
        val note: JsonObject,
        @SerialName("raw_transcript") val rawTranscript: String,
    ) : ServerMessage()

    @Serializable
    @SerialName("error")
    data class ErrorMessage(
        val code: String,
        val message: String,
    ) : ServerMessage()
}

val ProtocolJson: Json = Json {
    ignoreUnknownKeys = true
    classDiscriminator = "type"
}
